include/dlaf/permutations/general/impl.h

```diff
@@ -191,119 +236,135 @@ auto whenAllReadOnlyTilesArray(Matrix<const T, D>& matrix) {
-
-        // Note: that was guaranteed to be contiguous on allocation
+  using dlaf::common::DataDescriptor;
+
+  const SizeType vec_size = sz_loc.get<orthogonal(C)>();
+  auto sendrecv_f =
+      [vec_size](comm::Communicator& comm, std::vector<int> send_counts, std::vector<int> send_displs,
+                 const std::vector<pika::shared_future<matrix::Tile<const T, D>>>& send_tiles_fut,
+                 std::vector<int> recv_counts, std::vector<int> recv_displs,
+                 const std::vector<matrix::Tile<T, D>>& recv_tiles) {
+        // Note: both guaranteed to be column-major on allocation
         const T* send_ptr = send_tiles_fut[0].get().ptr();
         T* recv_ptr = recv_tiles[0].ptr();
 
-        // send displacements
-        std::exclusive_scan(send_counts.begin(), send_counts.end(), send_displs.begin(), 0);
+        const SizeType send_ld = send_tiles_fut[0].get().ld();
+        const SizeType recv_ld = recv_tiles[0].ld();
+
+        const SizeType send_perm_stride = (C == Coord::Col) ? send_ld : 1;
+        const SizeType recv_perm_stride = (C == Coord::Col) ? recv_ld : 1;
 
-        // recv displacements
+        // cumulative sum for computing rank data displacements in packed vectors
+        std::exclusive_scan(send_counts.begin(), send_counts.end(), send_displs.begin(), 0);
         std::exclusive_scan(recv_counts.begin(), recv_counts.end(), recv_displs.begin(), 0);
 
-        // All-to-All communication
-        DLAF_MPI_CHECK_ERROR(MPI_Ialltoallv(send_ptr, send_counts.data(), send_displs.data(), dtype,
-                                            recv_ptr, recv_counts.data(), recv_displs.data(), dtype,
-                                            comm, req));
+        std::vector<ex::unique_any_sender<>> all_comms;
+        all_comms.reserve(to_sizet(comm.size() - 1) * 2);
+        const comm::IndexT_MPI rank = comm.rank();
+        for (comm::IndexT_MPI rank_partner = 0; rank_partner < comm.size(); ++rank_partner) {
+          if (rank == rank_partner)
+            continue;
+
+          const auto rank_partner_index = to_sizet(rank_partner);
+
+          if (send_counts[rank_partner_index])
+            all_comms.push_back(
+                ex::just() | dlaf::comm::internal::transformMPI([=](MPI_Request* req) {
+                  const SizeType nperms = send_counts[rank_partner_index];
+                  auto message = dlaf::comm::make_message(
+                      DataDescriptor<const T>(send_ptr +
+                                                  send_displs[rank_partner_index] * send_perm_stride,
+                                              C == Coord::Col ? nperms : vec_size,
+                                              C == Coord::Col ? vec_size : nperms, send_ld));
+
+                  DLAF_MPI_CHECK_ERROR(MPI_Isend(message.data(), message.count(), message.mpi_type(),
+                                                 rank_partner, 0, comm, req));
```

Comment:
note-to-self: use a better TAG, in order to not rely on other constraints for avoiding mismatched communication.

@rasolca suggests to use `index_split` (wrt tridiagonal solver)